components/script/dom/window.rs

```diff
@@ -137,10 +137,15 @@ use webrender_api::{DocumentId, ExternalScrollId, RenderApiSender};
 use webvr_traits::WebVRMsg;
 
 /// Current state of the window object
-#[derive(Clone, Copy, Debug, JSTraceable, MallocSizeOf, PartialEq)]
+#[derive(JSTraceable, MallocSizeOf)]
+#[must_root]
 enum WindowState {
-    Alive,
-    Zombie, // Pipeline is closed, but the window hasn't been GCed yet.
+    /// A window with a browsing-context.
+    Alive(WindowStateData),
+    /// A window without a browsing-context,
+    /// keeping only its associated document, since it should always be kept.
+    /// See https://html.spec.whatwg.org/multipage/#concept-document-window
+    Zombie(MutNullableDom<Document>),
```

Comment:
I added this because this test started failing: `/html/browsers/the-window-object/document-attribute.window.html`, because if you do `frame.remove();` the browsing-context is discarded, and in servo the pipeline is closed, however the window/document are still potentially reachable from script.